src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1126,6 +1126,174 @@ object Desugar {
+
+              // get the length of the expression
+              length = in.Length(copiedVar)(rangeSrc)
+              cond = in.LessCmp(indexLeft, length)(rangeSrc)
+
+              // this invariant states that the index is between 0 and the length of the array/slice (both inclusive)
+              indexBoundsInv = in.ExprAssertion(in.And(
+                in.AtMostCmp(in.IntLit(0)(rangeSrc), indexLeft)(rangeSrc),
+                in.AtMostCmp(indexLeft, length)(rangeSrc))(rangeSrc))(rangeSrc)
+
+              // standard for loop desugaring
+              (dInvPre, dInv) <- prelude(sequence(spec.invariants map assertionD(ctx)))
+              (dTerPre, dTer) <- prelude(option(spec.terminationMeasure map terminationMeasureD(ctx)))
+
+              dBody = blockD(ctx)(body)
+
+              continueLabelName = nm.continueLabel(n, info)
+              continueLoopLabelProxy = in.LabelProxy(continueLabelName)(src)
+              continueLoopLabel = in.Label(continueLoopLabelProxy)(src)
+
+              breakLabelName = nm.breakLabel(n, info)
+              breakLoopLabelProxy = in.LabelProxy(breakLabelName)(src)
+              _ <- declare(breakLoopLabelProxy)
+              breakLoopLabel = in.Label(breakLoopLabelProxy)(src)
+
+              wh = if (shorts.length == 2) {
+                // here we know we have the value variable which we create
+                val valueSrc = meta(shorts(1))
+                val valueLeft = assignableVarD(ctx)(shorts(1))
+                val valueVar = in.Assignee.Var(valueLeft)
+                val valueAss = singleAss(valueVar, in.IndexedExp(copiedVar, in.IntLit(0)(valueSrc), typ)(valueSrc))(valueSrc)
+                val updateValue = in.If(cond, singleAss(valueVar, in.IndexedExp(copiedVar, indexLeft, typ)(valueSrc))(valueSrc), in.Seqn(Vector())(valueSrc))(valueSrc)
+
+                // we also need an invariant that states that
+                // for index i and value j and range expression x
+                // invariant 0 <= i && i < len(x) ==> j == x[i]
+                val indexValueEq = in.Implication(
+                  in.And(
+                    in.AtMostCmp(in.IntLit(0)(indexSrc), indexLeft)(indexSrc),
+                    in.LessCmp(indexLeft, length)(indexSrc))(indexSrc),
+                  in.ExprAssertion(in.EqCmp(in.IndexedExp(copiedVar, indexLeft, typ)(rangeExpSrc), valueLeft)(rangeExpSrc))(rangeExpSrc)
+              )(rangeSrc)
+                in.Block(
+                  Vector(valueLeft.asInstanceOf[in.LocalVar]),
+                  Vector(copyAss, indexAss, valueAss) ++ dInvPre ++ dTerPre ++ Vector(
+                    in.While(cond, dInv ++ Vector(indexBoundsInv, indexValueEq), dTer, in.Block(Vector(continueLoopLabelProxy),
+                      Vector(dBody, continueLoopLabel, incrIndex, updateValue) ++ dInvPre ++ dTerPre
+                    )(src))(src), breakLoopLabel
+                  )
+                )(src)
+              }
+              else {
```

Comment:
looks awkward, usually you see `} else {` in the same line

Reply:
Additionally, here you rely on the implicit assumption that the length of `shorts` is either 1 or 2. If for some reason, that assumption is violated (e.g., len(shorts) = 3), the code will still pass instead of reporting what is obviously an erroneous situtaion. Could we add a third case that reports a violation here?

Reply:
I didn't include it because I have handled that case in the type checker. Is there a situation where I shouldn't depend on the type checker to catch that case?